ui/src/external/monaco.flux.completions.ts

```diff
@@ -13,6 +13,83 @@ const m2p = new MonacoToProtocolConverter(),
   p2m = new ProtocolToMonacoConverter()
 
 export function registerCompletion(monaco: MonacoType, server: LSPServer) {
+  monaco.languages.registerDocumentSymbolProvider(FLUXLANGID, {
+    provideDocumentSymbols: async (model, _token) => {
+      try {
+        const uri = model.uri.toString()
+        const symbols = await server.symbols(uri)
+
+        return p2m.asSymbolInformations(symbols)
+      } catch (e) {
+        console.error(e)
+      }
+    },
+  })
+
+  monaco.languages.registerFoldingRangeProvider(FLUXLANGID, {
+    provideFoldingRanges: async (model, _context, _token) => {
+      try {
+        const ranges = await server.foldingRanges(model.uri.toString())
+
+        return p2m.asFoldingRanges(ranges)
+      } catch (e) {
+        // console.error(e)
+      }
+
+      return []
+    },
+  })
+
+  monaco.languages.registerDefinitionProvider(FLUXLANGID, {
+    provideDefinition: async (model, position, _token) => {
+      const uri = model.uri.toString()
+      const pos = m2p.asPosition(position.lineNumber, position.column)
+
+      try {
+        const loc = await server.definition(uri, pos)
+
+        return p2m.asLocation(loc)
```

Comment:
why doesn't this need the same guard as line 68?

Reply:
asLocation takes an optional type so it should handle a null value, whereas 68 we are calling map on a value that could be null.